fs/vfs/fs_timerfd.c

```diff
@@ -0,0 +1,824 @@
+    {
+      if (filep->f_oflags & O_NONBLOCK)
+        {
+          nxsem_post(&dev->exclsem);
+          return -EAGAIN;
+        }
+
+      timerfd_waiter_sem_t sem;
+      nxsem_init(&sem.sem, 0, 0);
+      nxsem_set_protocol(&sem.sem, SEM_PRIO_NONE);
+
+      do
+        {
+          ret = timerfd_blocking_io(dev, &sem, &dev->rdsems);
+          if (ret < 0)
+            {
+              nxsem_destroy(&sem.sem);
+              return ret;
+            }
+        }
+      while (timerfd_get_counter(dev) == 0);
+
+      nxsem_destroy(&sem.sem);
+    }
+
+  /* Device ready for read.  Ensure that interrupts are disabled and we
+   * do not lose counts if expiration occurs after read, but before setting
+   * counter to zero
+   */
+
+  intflags = enter_critical_section();
+
+  *(FAR timerfd_t *)buffer = dev->counter;
+  dev->counter = 0;
+
+  leave_critical_section(intflags);
+
+  nxsem_post(&dev->exclsem);
+  return sizeof(timerfd_t);
+}
+
+#ifdef CONFIG_TIMER_FD_POLL
+static int timerfd_poll(FAR struct file *filep, FAR struct pollfd *fds,
+                        bool setup)
+{
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct timerfd_priv_s *dev = inode->i_private;
+  int ret;
+  int i;
+
+  ret = nxsem_wait(&dev->exclsem);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  ret = OK;
+
+  if (!setup)
+    {
+      /* This is a request to tear down the poll. */
+
+      FAR struct pollfd **slot = (FAR struct pollfd **)fds->priv;
+
+      /* Remove all memory of the poll setup */
+
+      *slot                = NULL;
+      fds->priv            = NULL;
+      goto errout;
+    }
+
+  /* This is a request to set up the poll. Find an available
+   * slot for the poll structure reference
+   */
+
+  for (i = 0; i < CONFIG_TIMER_FD_NPOLLWAITERS; i++)
+    {
+      /* Find an available slot */
+
+      if (!dev->fds[i])
+        {
+          /* Bind the poll structure and this slot */
+
+          dev->fds[i] = fds;
+          fds->priv   = &dev->fds[i];
+          break;
+        }
+    }
+
+  if (i >= CONFIG_TIMER_FD_NPOLLWAITERS)
+    {
+      fds->priv = NULL;
+      ret       = -EBUSY;
+      goto errout;
+    }
+
+  /* Notify the POLLIN event if the counter is not zero */
+
+  if (timerfd_get_counter(dev) > 0)
+    {
+      timerfd_pollnotify(dev, POLLIN);
+    }
+
+errout:
+  nxsem_post(&dev->exclsem);
+  return ret;
+}
+#endif
+
+static void timerfd_timeout_work(FAR void *arg)
+{
+  FAR struct timerfd_priv_s *dev = (FAR struct timerfd_priv_s *)arg;
+  int ret;
+
+  ret = nxsem_wait_uninterruptible(&dev->exclsem);
+  if (ret < 0)
+    {
+      wd_cancel(&dev->wdog);
+      return;
+    }
+
+#ifdef CONFIG_TIMER_FD_POLL
+  /* Notify all poll/select waiters */
+
+  timerfd_pollnotify(dev, POLLIN);
+#endif
+
+  /* Notify all of the waiting readers */
+
+  timerfd_waiter_sem_t *cur_sem = dev->rdsems;
+  while (cur_sem != NULL)
+    {
+      nxsem_post(&cur_sem->sem);
+      cur_sem = cur_sem->next;
+    }
+
+  dev->rdsems = NULL;
```

Comment:
remove, let's timerfd_read restore the list

Reply:
This is copied from `eventfd_do_write`. Should we remove same line there as well?

Reply:
Reviewed the logic. Since all the waiters are notified we can clear `dev->rdsems = NULL;` here and omit `slist` removal logic from the `timerfd_read`. This simplifies code a bit.

Reply:
Yes, you are right!